adapter/internal/operator/controllers/cp/application_controller.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/wso2/apk/adapter/internal/loggers"
+	"github.com/wso2/apk/adapter/pkg/logging"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	cpv1alpha1 "github.com/wso2/apk/adapter/internal/operator/apis/cp/v1alpha1"
+	"github.com/wso2/apk/adapter/pkg/discovery/api/wso2/discovery/subscription"
+)
+
+// ApplicationReconciler reconciles a Application object
+type ApplicationReconciler struct {
+	client           client.Client
+	Scheme           *runtime.Scheme
+	applicationCache applicationCache
+}
+
+type applicationCache map[types.NamespacedName]*cpv1alpha1.Application
+
+// NewApplicationController creates a new Application controller instance.
+func NewApplicationController(mgr manager.Manager) error {
+	r := &ApplicationReconciler{
+		client:           mgr.GetClient(),
+		applicationCache: applicationCache{},
+	}
+	c, err := controller.New("Application", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		loggers.LoggerAPKOperator.ErrorC(logging.ErrorDetails{
+			Message:   fmt.Sprint("Error creating Application controller. ", err.Error()),
+			Severity:  logging.BLOCKER,
+			ErrorCode: 5001,
+		})
+		return err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &cpv1alpha1.Application{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		loggers.LoggerAPKOperator.ErrorC(logging.ErrorDetails{
+			Message:   fmt.Sprint("Error watching Application resources. ", err.Error()),
+			Severity:  logging.BLOCKER,
+			ErrorCode: 5002,
+		})
+		return err
+	}
+
+	loggers.LoggerAPKOperator.Info("Application Controller successfully started. Watching Application Objects...")
+	return nil
+}
+
+//+kubebuilder:rbac:groups=cp.wso2.com,resources=applications,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cp.wso2.com,resources=applications/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=cp.wso2.com,resources=applications/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// Modify the Reconcile function to compare the state specified by
+// the Application object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (r *ApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Comment:
please use meaningful names. fix it in all changes
```suggestion
func (applicationReconciler *ApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```